Clarify "The QUIC Handshake" (#387)
quicwg/ops-drafts: draft-ietf-quic-manageability.md

The content of Initial packets is encrypted using Initial Secrets, which
are derived from a per-version constant and the client's destination connection
ID; they are therefore observable by any on-path device that knows the
per-version constant and considered visible in this illustration. The content
of QUIC Handshake packets are encrypted using keys established during the
initial handshake exchange, and are therefore not visible.

Initial, Handshake, and 1-RTT packets belong to cryptographic and transport
contexts. The Client Completion {{fig-init-complete}} and the

> Server Completion {{fig-hs-complete}} flights conclude these first two
> contexts, by sending final acknowledgments and CRYPTO frames.

Reviewer: ```suggestion
Server Completion {{fig-hs-complete}} flights conclude the Initial and Handshake
contexts, by sending final acknowledgments and CRYPTO frames.
```

If you look at this, it's not a great match to reality.  The Initial context concludes at the server when it sends its first Handshake packet.  Is this "lies to children" and therefore acceptable?

Author: Yeah, this was my best effort at rewriting what was there to be clearer and more accurate, but it's still a bit misleading or possibly just not helpful.

I'm not sure if the intent is to detail when no more packets from an encryption level will be sent, because the keys have been discarded?

~~~~~
+----------------------------------------------------------+
| UDP header (source and destination UDP ports)            |
+----------------------------------------------------------+
| QUIC long header (type = Initial, Version, DCID, SCID) (Length)
+----------------------------------------------------------+  |
| QUIC Frames (ie: CRYPTO, PADDING, PING)                  |  |
+----------------------------------------------------------+  |

> | TLS Client Hello (incl. TLS SNI)                         |  |

Reviewer 2: does the TLS Client Hello not live within a QUIC frame? this makes it look otherwise in my reading.

Author: Thanks, I updated it to be clearer.  I also reverted part of my change and just put "Example" instead of the former "Typical", because the number of ways a client hello can be sequenced is extraordinarily large.

> completed.

Author: I'm actually not sure how one determines the handshake is complete from the server completion.  How does one tell the difference between the server sending 0.5RTT and 1RTT containing HANDSHAKE_DONE?

Reviewer: Maybe this is worth a note:
> The Server Completion flight might not be observable as it is indistinguishable from packets the server sends after the Server Initial flight.

{: #fig-client-initial-0rtt
   title="Typical 0-RTT Client Initial datagram pattern"}

When a 0-RTT packet is coalesced with an Initial packet, the datagram

> will be padded to 1200 byes. Additional datagrams containing only 0-RTT

Reviewer: Note: This is also "lies to children".  If a Handshake packet is coalesced with these (which can happen, depending on how the client works; ours will do this in some circumstances), then the datagram doesn't need to be padded.

Author: It doesn't?  I'm missing that in: https://www.rfc-editor.org/rfc/rfc9000.html#name-initial-datagram-size

I agree that in practice, if the client has Handshake keys but hasn't used them(ie: it's sending an Initial ACK before the Handshake packet), then there's no reason to pad the datagram, but I can't find text which makes an exception for that.

Reviewer: You know, I don't think that this is in the spec.  It's not a condition that the spec considers.  So maybe I can take that back.

It's what our implementation does though.  There is no point padding if you are sending Handshake, because the Handshake packet will unblock the server anti-amplification limit and you know that the server can process the Handshake packet (the only reason to send Handshake in this case is to acknowledge a Handshake packet).

Some implementations won't even send Initial in this case because they drop the keys.  We have this extra wrinkle that allows us to acknowledge Initial packets as a final gasp before dropping the keys.  It's a courtesy only and not reliable, but it means that we send Initial + Handshake + 0-RTT all in the same datagram.  AND that datagram won't be padded, even though it contains an Initial.

So I'm totally OK with this text remaining.  I certainly don't want to have to explain all of that.

Author: The 'courtesy ACKs' are actually useful because they provide an extra RTT sample early in the connection, so thanks for your courtesy :)
